### PR TITLE
[3.x] feat: Permissions Management

### PR DESCRIPTION
**Pull request by pboivin**
Redo of https://github.com/area17/twill/pull/1049 to move the source branch within the `area17/twill` repository.

A PR continuing the good work of #994. The original branch is being manually rebased on 2.x with a focus on maximizing backward compatibility. This is a work in progress :)

<br>

### Install the `area17/twill` package from the feature branch

```sh
composer require area17/twill:dev-feature/permissions-management
```

You may get the following error from composer: `Your requirements could not be resolved to an installable set of packages.`  This is a known issue (https://github.com/area17/twill/issues/1108). You can remove the `composer.lock` file and try again.

### Setup Twill

Edit your `.env` file, then run:

```sh
php artisan twill:install
```

### Create a test module

```sh
php artisan twill:make:module posts
```

Then, add the module to your `routes/admin.php` and `config/twill-navigation.php`

### Enable the feature

Add the following configuration to `config/twill.php`:

```
    'enabled' => [
        'permissions-management' => true,
    ],

    /*
    |--------------------------------------------------------------------------
    | Twill permissions
    |--------------------------------------------------------------------------
    |
    | This allows you to set up the level of management for the permissions.
    |
    | Three levels exist:
    | - role: this manages the access of modules based on the role of a user
    | - roleGroup: this manages the access of modules for a group of users, in addition to `role`
    | - roleGroupItem: this manages the permission of items within a module, in addition to `roleGroup`
    |
    */
    'permissions' => [
        'level' => 'role', // Define the permissions level you need
        'modules' => ['posts'], // Add all modules for which permissions should apply
    ],
```

### Run the feature's migrations

```sh
php artisan migrate
```

<br>

And you're done! You will find the permissions settings for Roles, Groups and Users in the `CMS Users` section.

### CODE REVIEW

**Comment by haringsrob** [comment]
I had a quick look/test and works great.

Is there a list of things to do/check?

**Comment by ifox** [comment]
> I had a quick look/test and works great.
> 
> Is there a list of things to do/check?

Great to hear, thanks @haringsrob!

First, it'd be great to merge 2.x into this branch. Then we should stress test this within an example app with the pattern you have implemented in #1376. Let's discuss this on our next call.

**Comment by haringsrob** [comment]
This one should now be ready to go into 3.x @pboivin @ifox

File: config/permissions.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    |
+    | Three levels exists
+    | - role: this manages the access of modules based on the role of a user.
+    | - roleGroup: this manages the access of modules for a group of users.
+    | - roleGroupItem: this manages the permission of items within a module.
+    |
+     */
+
+    'level' => 'role',
+    'modules' => [], // List of the modules the application have to manage access
```
Comment (haringsrob) [inline config/permissions.php:20]: @pboivin should this apply automatically to all if not set?

Comment (pboivin) [reply config/permissions.php:20]: Refreshing my memory on how roles are converted from the legacy system : [1](https://github.com/area17/twill/pull/1138/files#diff-039f5870c33f40d39f43a227f7e47a588b2a91c6d6459c85e0f44381fac8bb71R53) [2](https://github.com/area17/twill/pull/1138/files#diff-f6c89eeb9d2a9e7ae3333696a19a70d9591d5f700ca4c6389ed09648e9baaef2R148)

Seems like making all modules "permissionable" by default would also be non-breaking when upgrading from the default 2.x setup 👍️

**Comment by haringsrob** [comment]
- [ ] I have to rever my merge as I created an infinite loop

File: views/partials/navigation/_global_navigation.blade.php [hunk 2]
```diff
@@ -3,21 +3,26 @@
         @if(!empty(config('twill-navigation')))
             <ul class="header__items">
                 @foreach(config('twill-navigation') as $global_navigation_key => $global_navigation_element)
-                    @can($global_navigation_element['can'] ?? 'list')
+                    @php
+                        $is_settings = $global_navigation_key === 'settings';
+                        $is_module = $is_settings || ($global_navigation_element['module'] ?? false);
+                        $gate = $is_settings ? 'edit-settings' : ($global_navigation_element['can'] ?? 'access-module-list');
+                    @endphp
+                    @unless ($is_module && Auth::user()->cannot($gate, $global_navigation_key))
```
Comment (haringsrob) [inline views/partials/navigation/_global_navigation.blade.php:11]: Loops here?

File: views/partials/navigation/_primary_navigation.blade.php [hunk 3]
```diff
@@ -16,15 +27,18 @@
         <div class="container">
             <ul class="nav__list">
                 @foreach($primaryNavElements as $primary_navigation_key => $primary_navigation_element)
-                    @can($primary_navigation_element['can'] ?? 'list')
+                    @php
+                        $gate = $primary_navigation_element['can'] ?? 'access-module-list';
+                    @endphp
+                    @unless(($primary_navigation_element['module'] ?? false) && Auth::user()->cannot($gate, $primary_navigation_key))
```
Comment (haringsrob) [inline views/partials/navigation/_primary_navigation.blade.php:33]: also check here.

File: views/partials/navigation/_secondary_navigation.blade.php [hunk 4]
```diff
@@ -3,15 +3,18 @@
         <div class="container">
             <ul class="navUnder__list">
                 @foreach(config('twill-navigation.'. $_global_active_navigation . '.primary_navigation.' . $_primary_active_navigation)['secondary_navigation'] as $secondary_navigation_key => $secondary_navigation_element)
-                    @can($secondary_navigation_element['can'] ?? 'list')
+                    @php
+                        $gate = $secondary_navigation_element['can'] ?? 'access-module-list';
+                    @endphp
+                    @unless(($secondary_navigation_element['module'] ?? false) && Auth::user()->cannot($gate, $secondary_navigation_key))
```
Comment (haringsrob) [inline views/partials/navigation/_secondary_navigation.blade.php:9]: and here

File: config/permissions.php (hunk 1 above)
Comment (haringsrob) [reply config/permissions.php:20]: I think it makes sense to keep it "to be configured". So I will leave as is.

**Comment by haringsrob** [comment]
I am merging this as it currently is fully functional. We might improve on it as we progress with 3.x

**Pull request merged**